server.go

```diff
@@ -586,44 +587,76 @@ func (s *Server) ReadMetricSocket(serverConn net.PacketConn, packetPool *sync.Po
 }
 
 // ReadTraceSocket listens for available packets to handle.
-func (s *Server) ReadTraceSocket(packetPool *sync.Pool) {
+func (s *Server) ReadTraceSocket(serverConn net.PacketConn, packetPool *sync.Pool) {
 	// TODO This is duplicated from ReadMetricSocket and feels like it could be it's
 	// own function?
-
-	if s.TraceAddr == nil {
-		log.WithField("s.TraceAddr", s.TraceAddr).Fatal("Cannot listen on nil trace address")
-	}
 	p := packetPool.Get().([]byte)
 	if len(p) == 0 {
 		log.WithField("len", len(p)).Fatal(
 			"packetPool making empty slices: trace_max_length_bytes must be >= 0")
 	}
 	packetPool.Put(p)
 
-	// if we want to use multiple readers, make reuseport a parameter, like ReadMetricSocket.
-	serverConn, err := NewSocket(s.TraceAddr, s.RcvbufBytes, false)
-	if err != nil {
-		// if any goroutine fails to create the socket, we can't really
-		// recover, so we just blow up
-		// this probably indicates a systemic issue, eg lack of
-		// SO_REUSEPORT support
-		log.WithError(err).Fatal("Error listening for UDP traces")
-	}
-	log.WithField("address", s.TraceAddr).Info("Listening for UDP traces")
-
 	for {
 		buf := packetPool.Get().([]byte)
 		n, _, err := serverConn.ReadFrom(buf)
 		if err != nil {
-			log.WithError(err).Error("Error reading from UDP trace socket")
-			continue
+			// In tests, the probably-best way to
+			// terminate this reader is to issue a shutdown and close the listening
+			// socket, which returns an error, so let's handle it here:
+			select {
+			case <-s.shutdown:
+				log.WithError(err).Info("Ignoring ReadFrom error while shutting down")
+				return
+			default:
+				log.WithError(err).Error("Error reading from UDP trace socket")
+				continue
+			}
 		}
 
 		s.HandleTracePacket(buf[:n])
 		packetPool.Put(buf)
 	}
 }
 
+// ReadTraceStream reads a streaming connection in framed wire
+// format. See package github.com/stripe/veneur/protocol for details.
+func (s *Server) ReadTraceStream(serverConn net.Conn) {
+	defer func() {
+		serverConn.Close()
+	}()
+	tags := []string{"ssf_format:framed"}
+	for {
+		msg, err := protocol.ReadSSF(serverConn)
+		if err != nil {
+			if err == io.EOF {
+				// Client hangup, close this
+				s.Statsd.Incr("frames.disconnects", nil, 1.0)
+				return
+			}
+			if protocol.IsFramingError(err) {
+				log.WithError(err).
+					WithField("remote", serverConn.RemoteAddr()).
+					Error("Frame error reading from SSF connection. Closing.")
+				s.Statsd.Incr("ssf.error_total",
+					append([]string{"packet_type:unknown", "reason:framing"}, tags...),
+					1.0)
+				return
+			}
+			// Non-frame errors means we can continue reading:
+			log.WithError(err).
+				WithField("remote", serverConn.RemoteAddr()).
+				Error("Error processing an SSF frame")
+			s.Statsd.Incr("ssf.error_total",
+				append([]string{"packet_type:unknown", "reason:processing"}, tags...),
+				1.0)
+			continue
+		}
+		s.Statsd.Incr("ssf.received_total", tags, .1)
```

Comment:
Can you add the same `TODO` on here too? I don't want to pollute this PR with "redesign packet metrics" but this creates 1 UDP packet for every SSF frame and gives us a low-grade form of write amplification. :(